### PR TITLE
Updated CI Pipeline with ephemeral EC2 runners

### PR DESCRIPTION
**Pull request by vicilliar**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
CI/CD Pipeline update

* **What is the current behavior?** (You can also link to an open issue here)
1 CI/CD pipeline for all tests, incomplete environments

* **What is the new behavior (if this is a feature change)?**
1) separates the 1 big github workflow into multiple (6 total) so they can be run independently
2) uses ephemeral AWS EC2 runners (with 50gb memory) to run tests instead of default github runners
3) introduces CUDA and ARM tests and test environments

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
The marqo-api-tests pull request will need to be approved for these workflows to be usable (since they call updated files there).

* **Other information**:

### CODE REVIEW

File: .github/workflows/arm64_local_os.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          #pip install -r requirements.txt
+          pip install tox
+          pip install flake8
+      
+      # TODO: linting here
+  
+      - name: Checkout marqo-api-tests repo
+        uses: actions/checkout@v3
+        with:
+          repository: vicilliar/marqo-api-tests
```
Comment (pandu-k) [inline .github/workflows/arm64_local_os.yml]: Can we change this to the Marqo repo?

File: .github/workflows/arm64_local_os.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        uses: machulav/ec2-github-runner@v2
+        with:
+          mode: start
+          github-token: ${{ secrets.GH_PERSONAL_ACCESS_TOKEN }}
+
+          # ARM Runner Image
+          ec2-image-id: ami-08815b73e51f336f5
+          ec2-instance-type: a1.xlarge
+          subnet-id: subnet-011602eb168d440da
+          security-group-id: sg-060f59c9d3c43f2c5
```
Comment (pandu-k) [inline .github/workflows/arm64_local_os.yml]: Can we hide these IDs (via secrets). Feels a bit odd exposing these like this

File: .github/workflows/workflow_update_proposal.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
+# Pipeline Update Proposal
```
Comment (pandu-k) [inline .github/workflows/workflow_update_proposal.md]: Can you remove this file add this as an issue on the internal repo?

File: Dockerfile [hunk 4]
```diff
@@ -20,14 +20,11 @@ COPY requirements.txt requirements.txt
 RUN pip3 install --no-cache-dir -r requirements.txt
```
Comment (pandu-k) [inline Dockerfile]: Can you please merge in mainline - these changes should already be corrected

File: requirements.txt [hunk 5]
```diff
@@ -9,15 +9,16 @@ tox
 # s2_inference:
 more_itertools
```
Comment (pandu-k) [inline requirements.txt]: Can you revert changes in this file? Changing requirements makes it liable to break builds

File: Dockerfile (hunk 4 above)
Comment (pandu-k) [reply Dockerfile]: And please revert changes

**Review by pandu-k — CHANGES_REQUESTED** [review]
1. Changes requested in comments
2. Also can we compose actions together: https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/ to share major chunks of code?

**Review by pandu-k — CHANGES_REQUESTED** [review]
1. Changes requested in comments
2. Also can we compose actions together: https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/ to share major chunks of code?

File: .github/workflows/workflow_update_proposal.md (hunk 3 above)
Comment (vicilliar) [reply .github/workflows/workflow_update_proposal.md]: Which one is our internal repo? marqoai/marqo?

File: Dockerfile (hunk 4 above)
Comment (vicilliar) [reply Dockerfile]: How does this work? Can I merge without a PR approval?

File: .github/workflows/workflow_update_proposal.md (hunk 3 above)
Comment (vicilliar) [reply .github/workflows/workflow_update_proposal.md]: Nvm. Done!

**Review by pandu-k — APPROVED** [review]

**Pull request merged**